### PR TITLE
Add one-sided/two-sided truncated distribution and cdf/icdf method for univariate symmetric distributions

### LINKED ISSUE #895: Support for TruncatedDistribution
Currently, we have one-side truncated normal and cauchy implemented. But it would be nice to add a general class, which uses `cdf` method, for other distributions. There is a [nice discussion here](https://github.com/probtorch/pytorch/pull/121) for context on relevant numerical issues.

cc @fritzo

### PR DESCRIPTION
**Pull request by fehiepsi**
Resolves #895, #894, #914. The implementation only supports symmetric distributions because we can leverage the symmetry to address numerical issues that are discussed at [probtorch](https://github.com/probtorch/pytorch/pull/121) for the lower bound > 5. The issue is cdf/icdf suffers from precision errors at the right tail (e.g. `Normal().cdf(6.) = 1.`). Thanks to the symmetry, we can transform the right tail to the left tail, which has better precision (e.g. `Normal().cdf(-6.) = 9.865896e-10`).

### TODO
- [x] Add student T cdf/icdf
- [x] Support truncated from above
- [x] Add tests for cdf/icdf
- [x] Add tests for TruncatedDistribution
- [x] Make pytree test pass

### CODE REVIEW

**Comment by fehiepsi** [comment]
@fritzo Could you help me review this PR? Pls consider this as a low priority PR. A nice thing is your `gof` helps me fix several mistakes in the `sample` method. ;)

File: numpyro/distributions/continuous.py [hunk 1]
```diff
@@ -1049,6 +1071,240 @@ def variance(self):
...
+        base_dist = base_cls.tree_unflatten(base_aux, base_flatten)
+        return cls(base_dist, high=high)
+
+
+class TwoSidedTruncatedDistribution(Distribution):
+    arg_constraints = {"low": constraints.dependent, "high": constraints.dependent}
+    reparametrized_params = ["low", "high"]
+
+    def __init__(self, base_dist, low=0., high=1., validate_args=None):
+        assert isinstance(base_dist, (Cauchy, Laplace, Logistic, Normal, StudentT))
```
Comment (fritzo) [inline numpyro/distributions/continuous.py:1210]: Consider declaring types in a class-level tuple
```py
class TwoSidedTrincatedDistribution(Distribution):
    supported_types = (Cauchy, Laplace, Logistic, Normal, StudentT)
    def __init__(...):
        assert isinstance(base_dist, self.supported_types)
        ...
```
That way users can extend with custom distributions
```py
class MySymmetricDistribution(Distribution):
    def cdf(self, value): ...
    def icdf(self, value): ...

TwoSidedTruncatedDistribution.supported_types += (MySymmetricDistribution,)
```

File: test/test_distributions.py [hunk 2]
```diff
@@ -519,6 +548,28 @@ def test_log_prob(jax_dist, sp_dist, params, prepend_shape, jit):
     assert_allclose(jit_fn(jax_dist.log_prob)(samples), expected, atol=1e-5)
 
 
+@pytest.mark.parametrize('jax_dist, sp_dist, params', CONTINUOUS)
+def test_cdf_and_icdf(jax_dist, sp_dist, params):
+    if not sp_dist:
+        pytest.skip('no corresponding scipy distn.')
```
Comment (fritzo) [inline test/test_distributions.py:554]: Are all the `.cdf()` and `.icdf()` methods actually tested? You might want to add some simple algebraic tests that are independent of scipy:
- `d.cdf(d.icdf(u)) = u` for `u ~ uniform(0,1)`
- `d.icdf(d.cdf(x)) = x` for `x ~ d`
- `d/dx d.cdf(x) = d.log_prob(x).exp()` for `x ~ d`
- `d/du d.icdf(u) = 1 / d.log_prob(d.icdf(u)).exp()` for `u ~ uniform(0,1)`

File: numpyro/distributions/continuous.py [hunk 3]
```diff
@@ -1049,6 +1071,240 @@ def variance(self):
...
+        loc = self.base_dist.loc
+        sign = jnp.where(loc >= self.low, 1., -1.)
+        return self.base_dist.cdf(loc - sign * (loc - self.low))
+
+    @lazy_property
+    def _tail_prob_at_high(self):
+        # if low < loc, returns cdf(high) = 1; otherwise returns 1 - cdf(high) = 0
+        loc = self.base_dist.loc
+        sign = jnp.where(loc >= self.low, 1., -1.)
+        return 0.5 * (1 + sign)
```
Comment (fritzo) [inline numpyro/distributions/continuous.py:1120]: why not simply
```py
return jnp.where(self.low < self.base_dist.loc, 1., 0.)
```

File: numpyro/distributions/continuous.py [hunk 4]
```diff
@@ -1049,6 +1071,240 @@ def variance(self):
...
+        # if low < loc, returns cdf(high); otherwise returns 1 - cdf(high)
+        loc = self.base_dist.loc
+        sign = jnp.where(loc >= self.low, 1., -1.)
+        return self.base_dist.cdf(loc - sign * (loc - self.high))
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+        u = random.uniform(key, sample_shape + self.batch_shape)
+
+        # NB: we use a more numerical formula for a symmetric base distribution
```
Comment (fritzo) [inline numpyro/distributions/continuous.py:1242]: nit: "more numerical" -> "more numerically stable"

**Review by fritzo — COMMENTED** [review]
Interface looks great!  I haven't checked math of `.cdf()`,`.icdf()` but I would feel more confident if we added some simple algebraic tests independent of SciPy, testing with a large grid of random values.

File: numpyro/distributions/continuous.py (hunk 3 above)
Comment (fehiepsi) [reply numpyro/distributions/continuous.py:1120]: yeah, thanks

File: test/test_distributions.py (hunk 2 above)
Comment (fehiepsi) [reply test/test_distributions.py:567]: Thanks, those tests are nice! I just added them.

> Are all the .cdf() and .icdf() methods actually tested?

Yes, the test exercises all implemented methods.

File: test/test_distributions.py [hunk 5]
```diff
@@ -519,6 +548,40 @@ def test_log_prob(jax_dist, sp_dist, params, prepend_shape, jit):
...
+@pytest.mark.parametrize('jax_dist, sp_dist, params', CONTINUOUS)
+def test_cdf_and_icdf(jax_dist, sp_dist, params):
+    d = jax_dist(*params)
+    if d.event_dim > 0:
+        pytest.skip('skip testing cdf/icdf methods of multivariate distributions')
+    samples = d.sample(key=random.PRNGKey(0), sample_shape=(100,))
+    quantiles = random.uniform(random.PRNGKey(1), (100,) + d.shape())
+    try:
+        if d.shape() == ():
+            rtol = 1e-3 if jax_dist is dist.StudentT else 1e-5
```
Comment (fehiepsi) [inline test/test_distributions.py:560]: `rtol=1e-5` fails for StudentT. I guess the precision of betainc/its grad is not good enough or samples got extreme values.

**Review by fritzo — APPROVED** [review]

**Pull request merged**